### PR TITLE
Allow reconciliations on asset, liability and equity accounts only

### PR DESCRIPTION
**Pull request by ylavoie**
- Validate proper categories
- Warn on non-contiguous series
- Add Force and Skip buttons to apply a suitable fix or skip a test completely

### CODE REVIEW

**Comment by ehuelsmann** [comment]
Do we reject the input if it's not reconciled on a bank account? If so, this belongs in the migrations, if not, it probably belongs elsewhere (like: "sanity checks"; a category we don't have yet)

**Comment by ylavoie** [comment]
This belongs in sanity checks and we should create such a category. We do not reject the reconciliation currently, even if it has no functional value or of only some transactions a reconciled in an expense or cogs account.

**Comment by einhverfr** [comment]
People do reconcile non-bank accounts, whether petty cash drawers or AR/AP accounts.

**Comment by einhverfr** [comment]
Also, I can imagine that one might have reserve payment accounts that might not be flagged for the payment dropdown.

I don't think we should tie reconciliation to what dropdown it is currently a member of.  I think that's tying the wrong concerns together.

**Comment by einhverfr** [comment]
It would make more sense to me if the check was (at least for now) on the category field, and we restricted reconciliations to asset, liability, and equity accounts.  

I am having trouble imagining why one would reconcile an income or expense account, though.

**Comment by einhverfr** [comment]
And that seems to address the original concern (reconciliation of cogs accounts for example)

**Comment by ylavoie** [comment]
- We should tie it to the proper categories, there is no value to reconcile expenses or cogs.
- With the new possibility for the user to force a migration even if the tests flagged cases, we should reports cases when we see non-continuous series of reconciliations or usage on non-conventional accounts and let the user decide.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 1]
```diff
@@ -693,7 +693,7 @@ sub upgrade {
     for my $check (_applicable_upgrade_tests($dbinfo)) {
         my $sth = $request->{dbh}->prepare($check->test_query);
         $sth->execute()
-            or die "Failed to execute pre-migration check " . $check->name;
+            or die "Failed to execute pre-migration check " . $check->name . ", " . $sth->errstr;
```
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: This explains more in detail what is happening. I think it is justified for developers and migrators.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 2]
```diff
@@ -740,67 +740,57 @@ verify_check => md5_hex($check->test_query),
 
     my $rows = [];
     while (my $row = $sth->fetchrow_hashref('NAME_lc')) {
-      my $id = $row->{$check->{id_column}};
+      my $count = 1+scalar(@$rows);
+      #TODO: A '' will fail with invalid input syntax in fix_tests - YL
```
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: When dealing with multiple keys, we may find cases when not all are present in user record. We should deal with that.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 3]
```diff
@@ -853,41 +843,46 @@ sub fix_tests{
...
+        $query = "UPDATE $table SET $setters WHERE $where "
+               . join(' AND ',map {"$_ = ? "} @{$check->id_columns});
     }
     my $sth = $dbh->prepare($query);
 
     for my $count (1 .. $request->{count}){
-        my $id = $request->{"id_$count"};
         my @values;
-        push @values, $id
+        push @values, split(',',@{$check->id_columns})
```
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: Perl split is prohibited. In view of the fact that those strings has to go through the hiddens, we need to find a better way.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 4]
```diff
@@ -853,41 +843,46 @@ sub fix_tests{
...
         }
-        push @values, $request->{"id_$count"}
+        push @values, split(',',$request->{"id_$count"})
            if ! $check->{insert};
 
-        $sth->execute(@values) ||
+        my $rv = $sth->execute(@values) ||
             $request->error($sth->errstr);
+        #TODO: We should abort gracefully
+        die "$rv rows affected instead of only 1"
```
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: That ain't nice.

File: lib/LedgerSMB/Scripts/setup.pm (hunk 3 above)
Comment (ylavoie) [inline lib/LedgerSMB/Scripts/setup.pm]: Fixed, ProhibitStringySplit prohibits strings as first argument on split and mandates using a regex

**Comment by einhverfr** [comment]
If we want to tie to proper categories, then it seems to me the correct category is to allow reconciliation on asset, liability, and equity accounts but deny them on income and expense accounts.  The problem with cogs is the checkbox really only affects dropdown membership and isn't a permanent category.  Moreover you *can* select cogs and ar_paid if you like.....

File: UI/lib/elements.html [hunk 5]
```diff
@@ -432,7 +432,7 @@
 
 <?lsmb BLOCK tooltip;
   IF element_data && element_data.id.defined() && element_data.msg.defined();
-     props="connectId:'" _ element_data.id _ "'";
+     props="connectId:'" _ element_data.id.replace('_','-') _ "'";
```
Comment (ehuelsmann) [inline UI/lib/elements.html]: this doesn't make sense. Why would an id have an underscore that needs replacement with a hyphen?

It feels very much like this is papering over another problem: If the 'id' is explicitly set, we definitely shouldn't be changing it.

**Comment by ehuelsmann** [comment]
9cdd1e4de8a7b395d725770c10b7c855d6356323 can and should be a PR of its own: It has nothing to do with reconciliations.

**Comment by ehuelsmann** [comment]
2857e09ddb63ce1fb82b73c7bca6a93b14648457 fixes a problem coming from a PR not related to reconciliations. Please PR separately.

**Comment by ehuelsmann** [comment]
c0de833a77cfacffe14738055cfa1befe7ee593d too is the follow-up to another PR  unrelated to reconciliations. Please PR separately too.

**Comment by ehuelsmann** [comment]
733f6bb554f27ee7ac716fb3718cde25911f11fb and 29c64aa8dfea3534c311487d4b54c4b5dfbc89f1 seem like they make a net-zero effect. Could you drop both from the PR? it would seem they make the history unnecessarily lengthy.

**Comment by ehuelsmann** [comment]
I think 036374adcd0de2805a54958a5549e681a6799267 makes sense, but what's the "multi-key" bit there?

File: UI/lib/elements.html (hunk 5 above)
Comment (ylavoie) [reply UI/lib/elements.html]: I'm not changing the id of the main item, Dojo take cares of normalizing underscores to hyphens, so we need to point the tooltip to the normalized widgetid.

Comment (ehuelsmann) [reply UI/lib/elements.html]: It does? I know LedgerSMB "normalizes" items with a "name=" with an underscore to have an "id=" with a hyphen. Isn't your problem caused by the underscore-to-hyphen-in-LedgerSMB-html-generator code instead of by Dojo?

**Comment by einhverfr** [comment]
As a general note, I really think we should enforce one PR per topic.  And that whitespace changes (blank line deletions etc) should be their own PR.  In this case I see at least 4 major PR-worthy topics (blank line deletions, changing behaviour on upgrade tests, SQL-Ledger upgrades, and the check on past reconciliations).  This makes review and discussion a whole lot harder.  PRs should be atomic.

I would be for letting it slide on this case only but let's keep things simple going forward.

File: UI/lib/elements.html (hunk 5 above)
Comment (ylavoie) [reply UI/lib/elements.html]: The problem wasn't Dojo but us, trying to get nicer CSS in our Template elements.

Comment (ehuelsmann) [reply UI/lib/elements.html]: Can you be more verbose? This remark is too coarse. I have no idea what you're referring to.

Comment (ylavoie) [reply UI/lib/elements.html]: I added a comment in elements.html where we "normalize" id but affect tooltips, for which we compensate in tooltip generation.

Comment (ehuelsmann) [reply UI/lib/elements.html]: Actually, no, we don't. The code you commented doesn't apply to all ids so we should not compensate all cases: inputs which themselves provide an explicit "id" value don't get this transcoding (only those ids which are derived from "name" attributes)

Comment (ylavoie) [reply UI/lib/elements.html]: Fixed, we now normalize the button and the related tooltip at the same point. I removed the Tooltip patch.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 6]
```diff
@@ -417,6 +461,98 @@ selectable_values => { chart_id => q{SELECT concat(accno,' -- ',description) AS
...
+);
+
+#=pod
+
+# This must be done before the acc_trans test, for they have duplicates
+# and need a unique key
+    push @tests, __PACKAGE__->new(
+        # Add a unique key to allow editing
+        # Make the test serially reusable and protect against triggers
+        test_query => "ALTER TABLE acc_trans DISABLE TRIGGER USER;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: You are aware that checks like these may be run multiple times during migration? I'm wondering if this is the best way to deal with this situation. Isn't the key ['trans_id', 'entry_id'] unique?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 7]
```diff
@@ -1133,6 +1269,50 @@ push @tests, __PACKAGE__->new(
...
+                             UNION SELECT a.id, n.name        FROM ar a JOIN customer n ON n.id = a.customer_id
+                             UNION SELECT a.id, n.name        FROM ap a JOIN vendor n   ON n.id = a.vendor_id
+                         ) xx ON xx.id = ac.trans_id
+                         JOIN chart ch ON (ac.chart_id = ch.id)
+                        WHERE ( ch.category NOT IN ( 'A', 'L', 'Q' ) 
+                             OR ch.link NOT LIKE '%paid' ) 
+                          AND ac.cleared IS NOT NULL 
+                          AND ac.approved
+                     ORDER BY accno, transdate, ac.id",
+      display_name => marktext('Reconciliations on non-bank accounts'),
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: While I agree that this could be a sensible data integrity check, I'm wondering: does this situation prevent running an acceptable migration?

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 8]
```diff
@@ -233,6 +262,21 @@ sub _get_tests {
...
 
+push @tests, __PACKAGE__->new(
+    test_query => "SELECT * FROM pg_settings
+                    WHERE name = 'client_encoding'
+                      AND setting <> 'UTF8'",
+    display_name => marktext('Client encoding must be UTF8'),
+    name => 'client_encoding',
+    display_cols => ['name','setting','unit','category','short_desc','extra_desc','context','vartype','source','min_val','max_val','enumvals','boot_val','reset_val','sourcefile','sourceline'],
+ instructions => marktext(
+                   'Please fix the client encoding'),
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: Doesn't the client negotiate its own encoding? I mean, DBD::Pg >3.3.0 will always want to open a UTF8 encoded connection. If we require that, then why does the client encoding get set incorrectly?

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 6 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Entry_id won't exists until migration is started. Tests are done before that.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 7 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Nope, the user can force the migration even with those still present.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 8 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: What does this have to do with the current PR?

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 7 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Or skip this test completely

**Comment by ehuelsmann** [comment]
This PR includes the "skip" button from #3351. Merging should wait until that's merged; it'll allow better review of the actual change here.

File: lib/LedgerSMB/Scripts/setup.pm [hunk 9]
```diff
@@ -725,6 +725,8 @@ sub _applicable_upgrade_tests {
                   LedgerSMB::Upgrade_Tests->get_tests;
 }
 
+my @checks;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Scripts/setup.pm]: This line makes me think this PR does not yet include the insights from #3351.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 8 above)
Comment (ehuelsmann) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Indeed: what does this check have to do with the current PR? I think it does not relate to reconciliation aft all. Please remove this extra check from this PR. We might want to move it to the sanity checks later.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 10]
```diff
@@ -415,6 +472,94 @@ database, or delete the offending rows through PgAdmin III or psql'),
...
+);
+
+#=pod
+
+# This must be done before the acc_trans test, for they have duplicates
+# and need a unique key
+    push @tests, __PACKAGE__->new(
+        # Add a unique key to allow editing
+        # Make the test serially reusable and protect against triggers
+        test_query => 'ALTER TABLE acc_trans DISABLE TRIGGER USER;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: While I understand the problem, I'm not too hot about this solution: it abuses the validation system to execute migration code. 

Can you indicate what happens when this code is run twice?

File: old/lib/LedgerSMB/DBObject/Reconciliation.pm [hunk 11]
```diff
@@ -386,7 +386,10 @@ sub get {
     $self->{account_info} = $ref;
     ($ref) = $self->call_dbmethod(funcname=>'reconciliation__get_cleared_balance',
                                 args => { chart_id => $ref->{id},
-                                          report_date => $self->{end_date}->clone->add_interval('month',-1) });
+                                          report_date => $self->{end_date}->clone
```
Comment (ehuelsmann) [inline old/lib/LedgerSMB/DBObject/Reconciliation.pm]: This has been merged yesterday (don't render the PR# of hand)

File: sql/modules/Reconciliation.sql [hunk 12]
```diff
@@ -30,11 +30,17 @@ BEGIN;
...
 $$
+DECLARE
+        result NUMERIC;
 BEGIN
-        UPDATE cr_report set submitted = true where id = in_report_id;
-        PERFORM reconciliation__save_set(in_report_id, in_line_ids);
-
-        RETURN FOUND;
+        result = reconciliation__check_balanced(in_report_id);
+        IF result = 0 THEN
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: Why are you checking balancedness *before* adding the new line ids to the report?

File: sql/modules/Reconciliation.sql [hunk 13]
```diff
@@ -93,6 +115,30 @@ $$Sets which lines of the report are cleared.$$;
 CREATE OR REPLACE FUNCTION reconciliation__delete_my_report(in_report_id int)
 RETURNS BOOL AS
 $$
+DECLARE
+    failed BOOL;
+BEGIN
+    SELECT approved INTO failed FROM cr_report
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: No need for the `failed` variable here (which is unused in the result):

`PERFORM * FROM cr_report ...`

File: sql/modules/Reconciliation.sql [hunk 14]
```diff
@@ -93,6 +115,30 @@ $$Sets which lines of the report are cleared.$$;
...
+    WHERE id = in_report_id AND approved;
+    IF FOUND THEN
+        RAISE EXCEPTION 'Cannot delete approved recon report';
+        RETURN FALSE;
+    END IF;
+
+    -- Make sure that transactions present on this report do not have a
+    -- cleared_on date or they won't ever be accessible anymore.
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: If "posting" the report sets the cleared date, and posted reports can't be deleted, how come we may fail this test?

File: sql/modules/Reconciliation.sql [hunk 15]
```diff
@@ -93,6 +115,30 @@ $$Sets which lines of the report are cleared.$$;
...
+    SELECT COUNT(*)>0 INTO failed
+    FROM cr_report_line rl
+    JOIN acc_trans ac ON rl.ledger_id = ac.entry_id
+    WHERE report_id = in_report_id
+      AND cleared_on IS NOT NULL;
+
+    IF failed THEN
+        RAISE EXCEPTION 'Transactions already cleared on this report';
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: But the report itself isn't POSTED?

File: sql/modules/Reconciliation.sql [hunk 16]
```diff
@@ -101,9 +147,10 @@ $$
                                     and approved IS NOT TRUE);
     DELETE FROM cr_report
      WHERE id = in_report_id AND entered_username = SESSION_USER
-           AND submitted IS NOT TRUE AND approved IS NOT TRUE
-    RETURNING TRUE;
-$$ LANGUAGE SQL SECURITY DEFINER;
+           AND submitted IS NOT TRUE AND approved IS NOT TRUE;
+    RETURN FOUND;
+END;
+$$ LANGUAGE PLPGSQL SECURITY DEFINER;
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: Why do you want to change the procedure to PLPGSQL? The two conditions I see can't occur, or can be integrated.

File: sql/modules/Reconciliation.sql [hunk 17]
```diff
@@ -166,20 +213,28 @@ disable the block_change_when_approved trigger on cr_report.$$;
...
+BEGIN
+    SELECT COALESCE(
+        (SELECT SUM(ac.amount) * CASE WHEN c.category in('A', 'E') THEN -1 ELSE 1 END
+            INTO result
+            FROM account c
+            JOIN acc_trans ac ON (ac.chart_id = c.id)
+        JOIN (      SELECT id FROM ar WHERE approved
+              UNION SELECT id FROM ap WHERE approved
+              UNION SELECT id FROM gl WHERE approved
+              ) g ON g.id = ac.trans_id
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: Can be replaced by a single SELECT on the `transaction` table.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 18]
```diff
@@ -426,6 +426,60 @@ database, or delete the offending rows through PgAdmin III or psql'),
...
+ instructions => marktext('Make sure all meta numbers are unique.'),
+      appname => 'ledgersmb',
+  min_version => '1.3',
+  max_version => '1.4'
+);
+
+#=pod
+
+# This must be done before the acc_trans test, for they have duplicates
+# and need a unique key
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: actually, this will *always* run before the acc_trans test, even if the acc_trans test's UI has already been shown and "corrective information" has been provided.

There's one problem here though: because SQL doesn't guarantee set-ordering, the code below doesn't necessarily generate the same unique values for the same rows as the ones already in the "uniquified" set (and thus in the corrective data).

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 19]
```diff
@@ -1188,6 +1242,51 @@ push @tests, __PACKAGE__->new(
   max_version => '3.0'
 );
 
+    push @tests, __PACKAGE__->new(
+        # Add a unique key to allow editing
+        # Make the test serially reusable and protect against triggers
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: This comment doesn't apply?

File: sql/modules/Reconciliation.sql (hunk 13 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: Fixed

File: sql/modules/Reconciliation.sql (hunk 15 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: Agreed, so we should never have cleared transactions in it.

File: sql/modules/Reconciliation.sql (hunk 16 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: FOUND requires PLPGSQL

File: sql/modules/Reconciliation.sql (hunk 17 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: Fixed

File: lib/LedgerSMB/Scripts/setup.pm (hunk 9 above)
Comment (ylavoie) [reply lib/LedgerSMB/Scripts/setup.pm]: Fixed

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 10 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: Fixed, moved into pre-tests.

File: sql/modules/Reconciliation.sql (hunk 12 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: The entries are already added and a submit is asked for an unbalanced report.

File: sql/modules/Reconciliation.sql (hunk 14 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: Submit 2 reports, Refuse the first, Fix it by adding a transaction that was in the 2nd and you have an approved transaction in 2 reports and you can't remove it from the 2nd.
Batch submits in migration make ithat situation easy to arise.

File: sql/modules/Reconciliation.sql [hunk 20]
```diff
@@ -30,11 +30,17 @@ BEGIN;
...
+        result NUMERIC;
 BEGIN
-        UPDATE cr_report set submitted = true where id = in_report_id;
-        PERFORM reconciliation__save_set(in_report_id, in_line_ids);
-
-        RETURN FOUND;
+        result = reconciliation__check_balanced(in_report_id);
+        IF result = 0 THEN
+            UPDATE cr_report set submitted = true where id = in_report_id;
+            PERFORM reconciliation__save_set(in_report_id, in_line_ids);
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: If they have already been added, then what does the line above do?

File: lib/LedgerSMB/Upgrade_Pre_Tests.pm [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+=cut
+
+sub _get_pre_tests {
+    my ($request) = @_;
+
+    my @pre_tests;
+
+    push @pre_tests, __PACKAGE__->new(
+        # Add a unique key to allow editing
+        test_query => 'ALTER TABLE acc_trans DROP COLUMN IF EXISTS lsmb_entry_id;
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Pre_Tests.pm]: Why is the module called "pre tests" ? I don't see a test in this row nor in the one below.

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 22]
```diff
@@ -1,6 +1,6 @@
 =head1 NAME
 
-LedgerSMB::Upgrade_Tests - Upgrade tests for LedgerSMB
+LedgerSMB::Upgrade_Pre_Tests - Upgrade pre-tests for LedgerSMB
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: You renamed this module, but its file is still called Upgrade_Tests.

File: sql/modules/Reconciliation.sql (hunk 20 above)
Comment (ehuelsmann) [reply sql/modules/Reconciliation.sql]: Let me rephrase: what guarantees are there that the `in_line_ids` is exactly the same set of records as there is currently already in the report?

Comment (ehuelsmann) [reply sql/modules/Reconciliation.sql]: In other words: Is there any guarantee we're marking as reconciled, the same lines as the ones you used to check that the report is balanced?

File: sql/modules/Reconciliation.sql (hunk 16 above)
Comment (ehuelsmann) [reply sql/modules/Reconciliation.sql]: So does `RAISE EXCEPTION`; my point is: do we really need to raise an exception? Or simply add conditions to the DELETE statements which make them "update nothing when the preconditions are not met"?

The function used to be defined to return FALSE on failure. You've made it raise an exception on failure. That's a major change in paradigm. Is that required, or can we do with a smaller change?

File: sql/modules/Reconciliation.sql [hunk 23]
```diff
@@ -166,20 +211,26 @@ disable the block_change_when_approved trigger on cr_report.$$;
...
+            FROM account c
+            JOIN acc_trans ac ON (ac.chart_id = c.id)
+            JOIN transactions t on t.id = ac.trans_id
+        WHERE c.id = $1 AND cleared
+          AND t.approved IS TRUE
+          AND ac.approved IS true
+          AND ac.transdate <= in_report_date
+        GROUP BY c.id, c.category)
+    ,0);
+    RETURN result;
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: why do we need to convert this query/stored procedure to plpgsql? We're not doing anything specific that requires collecting the result into `result`.

File: sql/modules/Reconciliation.sql [hunk 24]
```diff
@@ -166,20 +211,26 @@ disable the block_change_when_approved trigger on cr_report.$$;
...
-          ) g ON g.id = ac.trans_id
-    WHERE c.id = $1 AND cleared
-      AND ac.approved IS true
-      AND ac.transdate <= in_report_date
-    GROUP BY c.id, c.category;
-$$ LANGUAGE sql;
+DECLARE
+        result NUMERIC;
+BEGIN
+    SELECT COALESCE(
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: This outer query is ineffective. The function result isn't being coalesced as it isn't stored in the `result` variable.

File: sql/modules/Reconciliation.sql (hunk 23 above)
Comment (ylavoie) [reply sql/modules/Reconciliation.sql]: Fixed

File: lib/LedgerSMB/Upgrade_Tests.pm [hunk 25]
```diff
@@ -1188,6 +1187,48 @@ push @tests, __PACKAGE__->new(
...
+                 ORDER BY trans_id, ac.id, accno, transdate),
+  display_name => marktext('Unneeded Reconciliations'),
+          name => 'reconciliation_on_unrelated_accounts',
+  display_cols => ['trans_id', 'id', 'memo', 'amount', 'description',
+                   'accno', 'account', 'link', 'category', 'cleared', 'approved'],
+       columns => ['cleared'],
+    id_columns => ['lsmb_entry_id'],
+  instructions => marktext(
+                   'Reconciliations should be on asset, liability or equity accounts only.<br>
+Please review the dates and categories in the original application if needed'),
```
Comment (ehuelsmann) [inline lib/LedgerSMB/Upgrade_Tests.pm]: why "if needed"? Is this a request to do so (and thus a validation failure) or not?

Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: They may be able to do all in LSMB or go back to SL to fix it.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 22 above)
Comment (ehuelsmann) [reply lib/LedgerSMB/Upgrade_Tests.pm]: No, I mean: You're changing the `=head1 NAME` section on  the wrong file.

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 25 above)
Comment (ehuelsmann) [reply lib/LedgerSMB/Upgrade_Tests.pm]: maybe instead of "if needed" -> "or click 'Skip' to continue" ?

File: sql/modules/Reconciliation.sql [hunk 26]
```diff
@@ -44,6 +50,22 @@ $$Submits a reconciliation report for approval.
...
+BEGIN
+      SELECT SUM(crl.our_balance) - cr.their_total INTO result
+        FROM cr_report_line crl
+        JOIN cr_report cr ON cr.id = crl.report_id
+       WHERE cr.id = in_report_id
+    GROUP BY cr.their_total;
+
+      RETURN COALESCE(result,0);
+END;
+$$ LANGUAGE PLPGSQL;
```
Comment (ehuelsmann) [inline sql/modules/Reconciliation.sql]: Same here. This can't be a `LANGUAGE SQL` function?

File: lib/LedgerSMB/Upgrade_Tests.pm (hunk 25 above)
Comment (ylavoie) [reply lib/LedgerSMB/Upgrade_Tests.pm]: I don't want to promote Skip. It is fixable and should be.
